packages/wonder-blocks-data/src/hooks/use-cached-effect.js

```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
+                    DataErrors.NotAllowed,
+                );
             }
-        },
-        [setMostRecentResult, onResultChanged, forceUpdate],
-    );
+            // We use our request fulfillment here so that in-flight
+            // requests are shared.
+            const request = RequestFulfillment.Default.fulfill(requestId, {
+                handler: interceptedHandler,
+            });
+
+            if (request === currentRequestRef.current?.request) {
+                // The request inflight is the same, so do nothing.
+                // NOTE: Perhaps if invoked via a refetch, we will want to
+                // override this behavior and force a new request?
+                return;
+            }
+
+            // Clear the last network result.
+            networkResultRef.current = null;
+
+            // Cancel the previous request.
+            currentRequestRef.current?.cancel();
+
+            // TODO(somewhatabstract, FEI-4276):
+            // Until our RequestFulfillment API supports cancelling/aborting, we
+            // will have to do it.
+            let cancel = false;
+
+            // NOTE: Our request fulfillment handles the error cases here.
+            // Catching shouldn't serve a purpose.
+            // eslint-disable-next-line promise/catch-or-return
+            request.then((result) => {
+                currentRequestRef.current = null;
+                if (cancel) {
+                    // We don't modify our result if the request was cancelled
+                    // as it means that this hook no longer cares about that old
+                    // request.
+                    return;
+                }
+
+                // Now we need to update the cache and notify or force a rerender.
+                setMostRecentResult(result);
+                networkResultRef.current = result;
+
+                if (onResultChanged != null) {
+                    // If we have a callback, call it to let our caller know we
+                    // got a result.
+                    onResultChanged(result);
+                } else {
+                    // If there's no callback, and this is using cache in some
+                    // capacity, just force a rerender.
+                    forceUpdate();
+                }
+                return; // Shut up eslint always-return rule.
+            });
+
+            currentRequestRef.current = {
+                requestId,
+                request,
+                cancel() {
+                    cancel = true;
+                    RequestFulfillment.Default.abort(requestId);
+                },
+            };
+        };
+        // We deliberately ignore the handler here because we want folks to use
+        // interceptor functions inline in props for simplicity. This is OK
+        // since changing the handler without changing the requestId doesn't
+        // really make sense - the same requestId should be handled the same as
+        // each other.
```

Comment:
This feels like it could admit subtle issues if the interceptor function captures variables from its closure.  Is this the existing behavior?

Reply:
Yes, it is.

Since interception is only used in fixtures and tests, I don't think that subtlety is an issue and we can call it out in docs if it becomes one.